---
title: "R Notebook"
output: html_notebook
---

The code in this notebook is used to take data from tables in the [CSBC/PS-ON Knowledge Portal Synapse project](https://www.synapse.org/csbcpson) and transform it into standardized and normalized tables for what I've dubbed the [CSBC/PS-ON Portal DB](https://www.synapse.org/#!Synapse:syn21498902/wiki/600349). A lot of the code is intended to be for one-time use, but some pieces might be repurposed for the ongoing data pipeline.

```{r}
library(synapser)
library(syndccutils)
library(tidyverse)
library(purrrogress)

source("../R/synapse_db.R")
source("../R/ror_client.R")

synapser::synLogin()
```

## Download

The `full_<type>_df` objects refer to "original" Tables and Views in the KP Project. The `cache = TRUE` argument creates a local feather file copy of the table, to save time from re-downloading. Switching back to `cache = FALSE` will force download of the latest version.

```{r}
use_local <- FALSE

full_synproject_df <- get_table_df("syn10142562", cache = use_local)
full_project_df <- get_table_df("syn21621978", cache = use_local)
full_dataset_df <- get_table_df("syn18488466", cache = use_local)
full_pub_df <- get_table_df("syn10923842", cache = use_local)
full_tool_df <- get_table_df("syn21488853", cache = use_local)
full_file_df <- get_table_df("syn9630847", cache = use_local)
```

## Normalize

Views for [`consortium`](https://www.synapse.org/#!Synapse:syn21630133) and [`theme`](https://www.synapse.org/#!Synapse:syn21630087) are ones that I created manually in the DB project. They're downloaded here and min-ified to enable creation of association/lookup tables.

### Consortium

```{r}
db_consortium_df <- get_table_df("syn21630133", cache = use_local)
```

```{r}
min_consortium_df <- db_consortium_df %>% 
    select(-createdOn, -createdBy, -modifiedOn, -modifiedBy, -currentVersion)

min_consortium_df
```

### Theme

```{r}
db_theme_df <- get_table_df("syn21630087", cache = use_local)
```

```{r}
min_theme_df <- db_theme_df %>% 
  select(-createdOn, -createdBy, -modifiedOn, -modifiedBy, -currentVersion)

min_theme_df
```

### Grant

#### Miminal Grant annotations

The min-ified `grant` table is intended to only include those attributes/fields that are specific to grants.

```{r}
# remove extraneous columns, clean up column names
min_grant_df <- full_synproject_df %>%
  filter(!is.na(grantNumber)) %>% 
  filter(!id %in% c("syn11857287", "syn7433743")) %>% 
  select(-type, -modifiedOn, -grantId, -grantName, -centerName,
         -teamMembersProfileId, -grantExternalSite, -fundingAgency) %>% 
  select(id, consortium, name, description = abstract, grantNumber, grantType,
         keyInvestigator = keyInvestigators, institution,
         website = centerSite, theme, synapseTeam = teamProfileId) %>% 
  mutate_all(str_trim)

# add the correct reference/FK for consortiumId
min_grant_df <- min_grant_df %>%
  left_join(select(min_consortium_df, 
                   consortiumId = id, 
                   consortium = displayName)) %>%
  select(-consortium)
```

#### Theme-Grant

```{r}
# pull out themes to create theme-grant association table 
# (many-to-many relationship), clean up theme names in the process
theme_grant_df <- min_grant_df %>% 
  select(grantId = id, theme) %>% 
  mutate(theme = str_replace(theme, "geneity/evol", "geneity; evol")) %>% 
  separate_rows(theme, sep = ";") %>%
  mutate(theme = str_to_lower(theme),
         theme = str_trim(theme)) %>%
  filter(!is.na(theme),
         theme != "") %>%
  mutate(theme = case_when(
    theme == "drug resistance/senstivity" ~ "drug resistance/sensitivity",
    theme == "tumor evolution" ~ "evolution",
    TRUE ~ theme
  ))

# add the correct reference/FK for themeId
theme_grant_df <- theme_grant_df %>% 
  # names converted to snake case for Synapse entity naming rules
  mutate(theme = snakecase::to_snake_case(theme)) %>% 
  left_join(select(min_theme_df, themeId = id, theme = name)) %>% 
  select(-theme)

# # initial table creation (synIDs need to be integers)
# theme_grant_syntable <- theme_grant_df %>%
#   mutate_all(~ str_replace(., "syn", "")) %>%
#   synBuildTable("theme_grant", "syn21498902", .) %>%
#   synStore()

# update/overwrite the table
theme_grant_syntable <- update_table("syn21639726", theme_grant_df)
```

#### Institution-Grant

```{r}
# pull out themes to create institution-grant association table 
# (many-to-many relationship), clean up institution names in the process
institution_grant_df <- min_grant_df %>% 
  select(grantId = id, name = institution) %>% 
  separate_rows(name, sep = ",") %>%
  separate_rows(name, sep = "/") %>%
  mutate(name = str_trim(name)) %>% 
  filter(!is.na(name)) %>% 
  mutate(fullName = case_when(
    name == "ASU" ~ "Arizona State University",
    name == "Harvard" ~ "Harvard University",
    name == "Moffitt Cancer Center & Research Institute" ~ "Moffitt Cancer Center",
    name == "MSKCC" ~ "Memorial Sloan Kettering Cancer Center",
    name == "MIT" ~ "Massachusetts Institute of Technology",
    name == "UC Berkeley" ~ "University of California, Berkeley",
    name == "UCSF" ~ "University of California, San Francisco",
    name == "UCSD" ~ "University of California, San Diego",
    name == "UCLA" ~ "University of California, Los Angeles",
    name == "UC Irvine" ~ "University of California, Irvine",
    name == "ISB" ~ "Institute for Systems Biology",
    name == "OHSU" ~ "Oregon Health & Science University",
    name == "UCB" ~ "University of California, Berkeley",
    name == "Vanderbilt" ~ "Vanderbilt University",
    name == "City of Hope" ~ "City Of Hope National Medical Center",
    name == "DFCI" ~ "Dana-Farber Cancer Institute",
    name == "Dartmouth" ~ "Dartmouth College",
    name == "Duke" ~ "Duke University",
    name == "Georgia Tech" ~ "Georgia Institute of Technology",
    name == "UTHSCSA" ~ "The University of Texas Health Science Center at San Antonio",
    name == "BIMDC" ~ "Beth Israel Deaconess Hospital",
    name == "MGH" ~ "Massachusetts General Hospital",
    name == "IUPUI" ~ "Indiana University â€“ Purdue University Indianapolis",
    name == "UMN" ~ "University of Minnesota",
    name == "UNC Chapel Hill" ~ "University of North Carolina at Chapel Hill",
    name == "NYUSOM" ~ "New York University School of Medicine",
    name == "Moffitt" ~ "Moffitt Cancer Center",
    name == "UVA" ~ "University of Virginia",
    name == "Institute of Computational Biology" ~ "University of Virginia",
    name == "UPenn" ~ "University of Pennsylvania",
    name == "USC" ~ "University of Southern California",
    name == "Houston Methodist Research Institute" ~ "Houston Methodist",
    TRUE ~ name
  )) %>% 
  distinct(grantId, fullName)

# retrieve standard metadata for institutions from Research Organization
# Registry (ROR) API
institution_df <- institution_grant_df %>% 
  distinct(fullName) %>% 
  mutate(org_data = map(fullName, get_org_ror)) %>% 
  unnest(org_data) %>% 
  mutate(rorId = str_replace(id, "https://ror.org/", "")) %>% 
  mutate(name = snakecase::to_upper_camel_case(name)) %>% 
  mutate(id = name) %>% 
  select(id, name, fullName, displayName = acronym, rorId) %>% 
  mutate(
    displayName = ifelse(str_length(displayName),
                         displayName,
                         fullName)
  ) %>% 
  select(-name) %>% 
  arrange(id)

# create institution entities in insitutions folder
institution_entities <- create_entities(institution_df, "syn21905841")

# manually create the view in Synapse and download (NOTE: need to add placeholder
# columns for addtiional annotations as well), then update the view
institution_synview <- update_view("syn21905891", institution_df)

# fetch the institution db table and update institution-grant
db_institution_df <- get_table_df("syn21905891")

# add the correct reference/FK for institutionId
institution_grant_df <- institution_grant_df %>% 
  left_join(select(db_institution_df, institutionId = id, fullName)) %>% 
  select(-fullName)

# # initial table creation (synIDs need to be integers)
# institution_grant_syntable <- institution_grant_df %>%
#   mutate_all(~ str_replace(., "syn", "")) %>%
#   synBuildTable("institution_grant", "syn21498902", .) %>%
#   synStore()

# update/overwrite the table
institution_grant_syntable <- update_table("syn21905912", institution_grant_df)
```

#### Person-Grant

```{r}
# pull out investigators to create person-grant association table 
# (many-to-many relationship), clean up investigator names in the process
person_grant_df <- min_grant_df %>% 
  select(grantId = id, person = keyInvestigator) %>% 
  separate_rows(person, sep = ",") %>% 
  mutate(person = str_trim(person)) %>% 
  filter(!is.na(person)) %>% 
  mutate(person = case_when(
    person == "Jill Mesirov" ~ "Jill P. Mesirov",
    person == "Douglas Lauffenburger" ~ "Douglas A. Lauffenburger",
    person == "Forest White" ~ "Forest M. White",
    TRUE ~ person
  )) %>% 
  mutate(role = "investigator")

# # initial table creation (synIDs need to be integers)
# person_grant_syntable <- person_grant_df %>%
#   mutate(grantId = str_replace(grantId, "syn", "")) %>%
#   synBuildTable("person_grant", "syn21498902", .) %>%
#   synStore()

# update/overwrite the table
person_grant_syntable <- update_table("syn21905932", person_grant_df)
```

#### Build/store Grant table (view)

```{r}
# select minimal grant columns and save table to Synapse
min_grant_df <- min_grant_df %>%
  select(id, consortiumId, name, description, grantNumber, grantType,
         website, synapseTeam) %>% 
  mutate(description = ifelse(str_detect(description, "^NA"), NA, description))

# # initial table creation (synIDs need to be integers)
# grant_syntable <- min_grant_df %>% 
#   left_join(select(full_synproject_df,
#                    id, createdOn, createdBy, modifiedOn, modifiedBy,
#                    currentVersion)) %>%
#   mutate_at(c("id", "consortiumId"), ~ str_replace(., "syn", "")) %>% 
#   synBuildTable("grant", "syn21498902", .) %>% 
#   synStore()

# update/overwrite the table
grant_syntable <- update_table("syn21639712", min_grant_df)
```

### Project

#### Minimal Project annotations

The min-ified `project` table is intended to only include those attributes/fields that are specific to projects.

```{r}
# remove extraneous columns, clean up column names
min_project_df <- full_project_df %>% 
  left_join(select(min_grant_df, grantId = id, grant = name)) %>% 
  mutate(id = str_c("proj", row_number(), sep = "_")) %>% 
  mutate(projectType = str_replace(projectNumber, " [0-9]", "")) %>% 
  select(id, grantId, name = projectName, description, 
         projectLabel = projectNumber, projectType,
         investigator = principalInvestigators) %>% 
  mutate(projectLabel = str_to_lower(str_replace(projectLabel, " ", "_"))) %>%
  mutate(id = str_c(grantId, projectLabel, sep = "_")) %>%
  mutate_all(str_trim)
```

#### Description-Project

```{r}
# pull out descriptions (which exceed column size limits for File Views) 
# to create description-project association table (one-to-one relationship)
description_project_df <- min_project_df %>% 
  distinct(grantId, projectLabel, description) %>%
  mutate(projectLabel = str_to_lower(str_replace(projectLabel, " ", "_"))) %>%
  mutate(name = str_c(grantId, projectLabel, sep = "_")) %>% 
  select(name, description)
```

#### Person-Project

```{r}
# pull out people (investigators) to create person-project association table 
# (many-to-many relationship... ?)
# NOTE: not actually using this for now
person_project_df <- min_project_df %>% 
  select(id, person = investigator) %>%
  separate_rows(person, sep = "(,| and )") %>% 
  mutate(person = str_trim(person)) %>%
  filter(!is.na(person))
```

#### Build/store Project table (view)

```{r}
project_df <- min_project_df %>%
  mutate(projectLabel = snakecase::to_title_case(projectLabel)) %>% 
  mutate(displayName = name) %>% 
  mutate(id = ifelse(is.na(id), str_c("proj", row_number(), sep = "_"), id)) %>% 
  select(id, fullName = name, displayName, projectLabel, projectType)

# create project entities in projects folder
project_entities <- create_entities(project_df, "syn21645004")

# manually create the view in Synapse and download (NOTE: need to add placeholder
# columns for addtiional annotations as well), then update the view
project_synview <- update_view("syn21645147", project_df)

# fetch the project db table for mapping project IDs below
db_project_df <- get_table_df("syn21645147")
```

#### Store association tables

```{r}
# add the correct reference/FK for projectId
description_project_df <- description_project_df %>% 
  left_join(select(db_project_df, projectId = id, name)) %>% 
  select(projectId, description)

# # initial table creation (synIDs need to be integers)
# description_project_syntable <- description_project_df %>% 
#   mutate_at(c("projectId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("description_project", "syn21498902", .) %>% 
#   synStore()

# update/overwrite the table
description_project_syntable <- update_table("syn21868407", description_project_df)
```

### Publication

#### Minimal Publication annotations

The min-ified `publication` table is intended to only include those attributes/fields that are specific to publications.

I'll create update this before datasets/tools to collect information about links.

```{r}
link_cols <- c(
  "GSE", "GSELink",
  "SRX", "SRXLink",
  "SRP", "SRPLink",
  "dbGaP", "dbGaPLink",
  "toolLink"
)

min_pub_df <- full_pub_df %>% 
  select(-consortium, -grantType, -grantName, -theme) %>%
  mutate(pubMedId = str_extract(pubMedUrl,
                                "[0-9].*")) %>%
  mutate(id = str_c("pmid_", pubMedId)) %>% 
  select(id, grantNumber, title = publicationTitle, journal,
         publicationYear, doi, pubMedUrl,
         keywords,
         author = authors,
         tissue, assay, tumorType,
         annotatedBy = `annotated.by`,
         one_of(link_cols)) %>%
  separate_rows(grantNumber, sep = ",") %>% 
  filter(grantNumber %in% min_grant_df$grantNumber) %>%
  left_join(select(min_grant_df, grantId = id, grantNumber)) %>% 
  select(id, grantId, grantNumber, title, journal,
         publicationYear, doi, pubMedUrl,
         keywords, author, tissue, assay, tumorType, annotatedBy,
         one_of(link_cols)) %>%
  select(-grantNumber) %>% 
  mutate_all(str_trim)
```

#### Link-Publication

```{r}
# pull out links to create link-publication association table 
# (one-to-many relationship from pub -> links â€” I think)
link_pub_df <- min_pub_df %>%   # select(link_cols) %>% 
  select(id, one_of(link_cols)) %>%
  gather(source, value, -id) %>%
  mutate(valueType = ifelse(str_detect(source, "Link"),
                            "url", "name")) %>%
  mutate(source = str_replace(source, "Link", "")) %>%
  pivot_wider(id_cols = c("id", "source"),
              names_from = valueType, values_from = value,
              values_fn = list(value = unique)) %>%
  arrange(id) %>%
  mutate(url = case_when(
    str_detect(url, "REVEALER") ~ "https://github.com/genepattern/REVEALER",
    str_detect(url, "www.ndexbio.org") ~ "http://www.ndexbio.org/",
    TRUE ~ url
  )) %>%
  mutate(source = case_when(
    str_detect(url, "github") ~ "GitHub",
    str_detect(url, "bitbucket") ~ "Bitbucket",
    str_detect(url, "readthedocs") ~ "ReadTheDocs",
    str_detect(url, "bioconductor") ~ "Bioconductor",
    source == "tool" ~ "Website",
    source == "GSE" ~ "GEO",
    source == "SRP" ~ "SRA",
    TRUE ~ source
  )) %>%
  filter(!is.na(url)) %>%
  separate_rows(name, sep = ",") %>%
  separate_rows(url, sep = ",") %>%
  filter((!is.na(name) & str_detect(url, name)) | is.na(name)) %>% 
  replace_na(list(name = "Tool")) %>% 
  rename(publication = id)

link_pub_df
```

#### Person-Publication

```{r}
# pull out people (authors) to create person-publication association table 
# (many-to-many relationship)
person_pub_df <- min_pub_df %>% 
  select(publication = id, person = author) %>% 
  separate_rows(person, sep = "(,| and )") %>% 
  mutate(person = str_trim(person)) %>% 
  mutate(role = "author") %>% 
  distinct() %>% 
  filter(!is.na(person))

person_pub_df
```

#### Grant-Publication

```{r}
# pull out grants to create grant-publication association table 
# (many-to-many relationship)
grant_pub_df <- min_pub_df %>% 
  select(publication = id, grantId) %>% 
  distinct() %>%
  filter(!is.na(grantId))

grant_pub_df
```

#### Assay-Publication

```{r}
# pull out assays to create assay-publication association table 
# (many-to-many relationship)
assay_pub_df <- min_pub_df %>% 
  select(publication = id, assay) %>% 
  separate_rows(assay, sep = ";") %>% 
  separate_rows(assay, sep = ",") %>% 
  mutate(assay = str_trim(assay)) %>% 
  filter(!is.na(assay)) %>% 
  mutate(assay = case_when(
    assay == "ATAC-seq" ~ "ATAC-Seq",
    str_detect(str_to_lower(assay), "cell migration assay") ~ "Cell Migration Assay",
    str_detect(str_to_lower(assay), "cell viabililty assay") ~ "Cell Viability Assay",
    str_detect(str_to_lower(assay), "computational modeling") ~ "Computational Modeling",
    str_detect(str_to_lower(assay), "computed tomography") ~ "Computated Tomography",
    str_detect(str_to_lower(assay), "expression array") ~ "Expression Array",
    str_detect(str_to_lower(assay), "flow cytometry") ~ "Flow Cytometry",
    str_detect(str_to_lower(assay), "fluorescence correlation spectroscopy") ~ "Fluorescence Correlation Spectroscopy",
    str_detect(str_to_lower(assay), "rna-seq") ~ "Whole Transcriptome Sequencing",
    str_detect(str_to_lower(assay), "rnaseq") ~ "Whole Transcriptome Sequencing",
    str_detect(str_to_lower(assay), "shrna") ~ "shRNA",
    str_detect(str_to_lower(assay), "single cell rna-sequencing") ~ "Single Cell RNA-Sequencing",
    str_detect(str_to_lower(assay), "imaging") ~ "Imaging",
    str_detect(str_to_lower(assay), "immunocytochemistry") ~ "Immunocytochemistry",
    str_detect(str_to_lower(assay), "mass cytometry") ~ "Mass Cytometry",
    str_detect(str_to_lower(assay), "mass spectrometry") ~ "Mass Spectrometry",
    str_detect(str_to_lower(assay), "mathematical modeling") ~ "Mathematical Modeling",
    str_detect(str_to_lower(assay), "microscopy") ~ "Microscopy",
    str_detect(str_to_lower(assay), "mouse model(s)*") ~ "Murine Model",
    str_detect(str_to_lower(assay), "murine model") ~ "Murine Model",
    str_detect(str_to_lower(assay), "not applicable") ~ "Not Applicable",
    str_detect(str_to_lower(assay), "organoid") ~ "Organoid",
    str_detect(str_to_lower(assay), "single cell rna-sequencing") ~ "Single Cell RNA-Sequencing",
    str_detect(str_to_lower(assay), "single-cell rna sequencing") ~ "Single Cell RNA-Sequencing",
    str_detect(str_to_lower(assay), "whole exome sequencing") ~ "Whole Exome Sequencing",
    str_detect(str_to_lower(assay), "whole genome sequencing") ~ "Whole Genome Sequencing",
    str_detect(str_to_lower(assay), "whole transcriptome sequencing") ~ "Whole Transcriptome Sequencing",
    TRUE ~ assay
  )) %>% 
  distinct() 

assay_pub_df
```

#### Tissue-Publication

```{r}
# pull out tissues to create tissue-publication association table 
# (many-to-many relationship)
tissue_pub_df <- min_pub_df %>% 
  select(publication = id, tissue) %>% 
  separate_rows(tissue, sep = ";") %>% 
  separate_rows(tissue, sep = ",") %>% 
  mutate(tissue = str_trim(tissue)) %>% 
  filter(!is.na(tissue)) %>% 
  mutate(tissue = case_when(
    str_detect(str_to_lower(tissue), "bone marrow") ~ "Bone Marrow",
    str_detect(str_to_lower(tissue), "colon") ~ "Colon",
    str_detect(str_to_lower(tissue), "nervous system") ~ "Nervous system",
    str_detect(str_to_lower(tissue), "not applicable") ~ "Not Applicable",
    str_detect(str_to_lower(tissue), "not appicable") ~ "Not Applicable",
    str_detect(str_to_lower(tissue), "ovary") ~ "Ovary",
    str_detect(str_to_lower(tissue), "small intestine") ~ "Small Intestine",
    TRUE ~ tissue
  )) %>% 
  distinct()

tissue_pub_df
```

#### TumorType-Publication

```{r}
# pull out tummor types to create tumortype-publication association 
# table (many-to-many relationship)
tumortype_pub_df <- min_pub_df %>% 
  select(publication = id, tumorType) %>% 
  separate_rows(tumorType, sep = ";") %>% 
  separate_rows(tumorType, sep = ",") %>% 
  mutate(tumorType = str_trim(tumorType)) %>% 
  filter(!is.na(tumorType)) %>% 
  mutate(tumorType = case_when(
    str_detect(str_to_lower(tumorType), "glioblastoma") ~ "Glioblastoma",
    str_detect(str_to_lower(tumorType), "intraductal carcinoma in situ of breast") ~ "Intraductal carcinoma in situ of Breast",
    str_detect(str_to_lower(tumorType), "malignant neoplasm of pancreas") ~ "Malignant Neoplasm of Pancreas",
    str_detect(str_to_lower(tumorType), "not applicable") ~ "Not Applicable",
    str_detect(str_to_lower(tumorType), "pancancer") ~ "Pan-cancer",
    TRUE ~ tumorType
  )) %>% 
  distinct()

tumortype_pub_df
```

#### Dataset-Publication

```{r}
# pull out datasets to create dataset-publication association table 
# (many-to-many relationship)
dataset_pub_df <- min_pub_df %>% 
  select(publication = id) %>% 
  left_join(link_pub_df) %>% 
  filter(!is.na(source), name != "Tool") %>% 
  mutate(dataset = str_to_lower(name)) %>%
  mutate(dataset = str_replace_all(dataset, "\\.", "_")) %>%
  select(publication, dataset) %>% 
  distinct() 
  
dataset_pub_df
```

#### Build/store Publication table (view)

```{r}
publication_df <- min_pub_df %>%
  select(id, title, journal, publicationYear, doi, keywords) %>% 
  distinct()

# create publication entities in publications folder
pub_entities <- create_entities(publication_df, "syn21645243")

# manually create the view in Synapse and download (NOTE: need to add placeholder
# columns for addtiional annotations as well), then update the view
pub_synview <- update_view("syn21650536", publication_df)

# fetch the project db table for mapping project IDs below
db_pub_df <- get_table_df("syn21650536")
```

#### Store association tables

```{r}
# add the correct reference/FK for publicationId
link_pub_df <- link_pub_df %>% 
  left_join(select(db_pub_df, publicationId = id, publication = name)) %>% 
  select(publicationId, source, name, url)

# # initial table creation (synIDs need to be integers)
# link_pub_syntable <- link_pub_df %>%
#   mutate_at(c("publicationId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("link_publication", "syn21498902", .) %>%
#   synStore()

# update/overwrite the table
link_pub_syntable <- update_table("syn21868571", link_pub_df)
```

```{r}
# add the correct reference/FK for publicationId
person_pub_df <- person_pub_df %>% 
  left_join(select(db_pub_df, publicationId = id, publication = name)) %>% 
  select(publicationId, person, role)

# # initial table creation (synIDs need to be integers)
# person_pub_syntable <- person_pub_df %>%
#   mutate_at(c("publicationId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("person_publication", "syn21498902", .) %>%
#   synStore()

person_pub_syntable <- update_table("syn21792842", person_pub_df)
```

```{r}
# add the correct reference/FK for publicationId
grant_pub_df <- grant_pub_df %>% 
  left_join(select(db_pub_df, publicationId = id, publication = name)) %>% 
  select(publicationId, grantId)

# # initial table creation (synIDs need to be integers)
# grant_pub_syntable <- grant_pub_df %>%
#   mutate_at(c("publicationId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("grant_publication", "syn21498902", .) %>%
#   synStore()

grant_pub_syntable <- update_table("syn21682660", grant_pub_df)
```


```{r}
# add the correct reference/FK for publicationId
assay_pub_df <- assay_pub_df %>% 
  left_join(select(db_pub_df, publicationId = id, publication = name)) %>% 
  select(publicationId, assay)

# # initial table creation (synIDs need to be integers)
# assay_pub_syntable <- assay_pub_df %>%
#   mutate_at(c("publicationId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("assay_publication", "syn21498902", .) %>%
#   synStore()

assay_pub_syntable <- update_table("syn21794722", assay_pub_df)
```

```{r}
# add the correct reference/FK for publicationId
tissue_pub_df <- tissue_pub_df %>% 
  left_join(select(db_pub_df, publicationId = id, publication = name)) %>% 
  select(publicationId, tissue)

# # initial table creation (synIDs need to be integers)
# tissue_pub_syntable <- tissue_pub_df %>%
#   mutate_at(c("publicationId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("tissue_publication", "syn21498902", .) %>%
#   synStore()

tissue_pub_syntable <- update_table("syn21795296", tissue_pub_df)
```

```{r}
# add the correct reference/FK for publicationId
tumortype_pub_df <- tumortype_pub_df %>% 
  left_join(select(db_pub_df, publicationId = id, publication = name)) %>% 
  select(publicationId, tumorType)

# # initial table creation (synIDs need to be integers)
# tumortype_pub_syntable <- tumortype_pub_df %>%
#   mutate_at(c("publicationId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("tumortype_publication", "syn21498902", .) %>%
#   synStore()

tumortype_pub_syntable <- update_table("syn21795742", tumortype_pub_df)
```

```{r}
# add the correct reference/FK for publicationId
dataset_pub_df <- dataset_pub_df %>% 
  left_join(select(db_pub_df, publicationId = id, publication = name)) %>% 
  select(publicationId, datasetId = dataset)

# # initial table creation (synIDs need to be integers)
# dataset_pub_syntable <- dataset_pub_df %>%
#   mutate_at(c("publicationId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("dataset_publication", "syn21498902", .) %>%
#   synStore()
```

### Dataset

#### Minimal Dataset annotations

The min-ified `dataset` table is intended to only include those attributes/fields that are specific to datasets

```{r}
# remove extraneous columns, clean up column names
min_dataset_df <- full_dataset_df %>%
  filter(is.dataset == "TRUE") %>% 
  select(-datasets, -tumorType, -species, -consortium, -is.dataset, -is.study,
         -createdOn, -createdBy, -modifiedOn, -modifiedBy, -etag,
         -grantType, -num.files, -featured, -dataset,
         -Theme, -studies, -centerName, -Title, -publicationTitle, 
         -theme, -datasetId, -experimentalStrategy) %>% 
  left_join(select(min_grant_df, grantId = id, grantName = name)) %>% 
  select(-grantName) %>% 
  select(id, grantId, name, displayName = portalDisplayName,
         description, overallDesign, externalLink, pubMedUrl = PubMed) %>% 
  mutate(synapseLocation = id,  id = name) %>% 
  mutate_all(str_trim) %>% 
  filter(str_detect(name, "GSE"))

# fetch the link_publication db table
db_link_pub <- get_table_df("syn21868571", cache = TRUE)
```

```{r}
# pull out publications to create publication-dataset association table 
# (many-to-many relationship)
# NOTE: only using this locally for now
pub_dataset_df <- min_dataset_df %>% 
  mutate(pubMedId = str_extract(pubMedUrl, "[0-9].*")) %>% 
  select(id, pubMedId)
  
pub_dataset_df <- full_join(pub_dataset_df, db_link_pub, 
                            by = c("id" = "name")) %>%
  left_join(select(db_pub_df, publicationId = id, pmid = name)) %>%
  filter(id != "Tool") %>%
  mutate(pmid = str_replace(pmid, "pmid_", "")) %>%
  mutate(pubMedId = ifelse(is.na(pubMedId), pmid, pubMedId)) %>% 
  mutate(name = id) %>%
  select(id, name, pubMedId) %>%
  distinct() %>%
  mutate(id = str_to_lower(id)) %>%
  mutate(id = str_replace_all(id, "\\.", "_")) %>%
  arrange(pubMedId) %>% 
  left_join(select(db_pub_df, publicationId = id, name) %>% 
              mutate(pubMedId = str_replace(name, "pmid_", "")) %>% 
              select(-name)
)

pub_dataset_df
```

#### Link-Dataset

```{r}
# pull out links to create link-dataset association table 
# (one-to-one, possibly one-to-many relationship for dataset -> links)
link_dataset_df <- pub_dataset_df %>% 
  left_join(db_link_pub) %>% 
  rowwise() %>% 
  mutate(
    url = ifelse(is.na(url) & str_detect(name, "GSE"),
                 paste0("https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=", name),
                 url)
  ) %>% 
  mutate(source = case_when(
    str_detect(name, "GSE") ~ "GEO",
    str_detect(name, "SRP") ~ "SRA",
    TRUE ~ source
  )) %>%
  select(dataset = id, source, name, url) %>% 
  distinct()

link_dataset_df
```

#### Description-Dataset

```{r}
# pull out descriptions (which exceed column size limits for File Views) 
# to create description-dataset association table (one-to-one relationship)
description_dataset_df <- min_dataset_df %>% 
  select(dataset = name, description) %>% 
  filter(!is.na(description)) %>% 
  mutate(dataset = str_to_lower(dataset)) %>%
  mutate(dataset = str_replace_all(dataset, "\\.", "_")) %>%
  distinct()

description_dataset_df
```

#### Build/store Dataset table (view)

```{r}
dataset_df <- pub_dataset_df %>%
  left_join(select(min_dataset_df, -id)) %>% 
  rename(fullName = displayName) %>% 
  select(id, displayName = name, fullName, 
         overallDesign, synapseLocation) %>% 
  group_by(id, displayName) %>% 
  slice(1) %>% 
  distinct()

# create dataset entities in datasets folder
dataset_entities <- create_entities(dataset_df, "syn21889510")

# manually create the view in Synapse and download (NOTE: need to add placeholder
# columns for addtiional annotations as well), then update the view
dataset_synview <- update_view("syn21889931", dataset_df)

# fetch the dataset db table for mapping dataset IDs below
db_dataset_df <- get_table_df("syn21889931")
```

#### Store association tables

```{r}
# add the correct reference/FK for datasetId
dataset_pub_df <- dataset_pub_df %>% 
  rename(dataset = datasetId) %>% 
  left_join(select(db_dataset_df, datasetId = id, dataset = name)) %>% 
  select(publicationId, datasetId)

dataset_pub_syntable <- update_table("syn21895676", dataset_pub_df)
```

```{r}
# add the correct reference/FK for datasetId
link_dataset_df <- link_dataset_df %>%
  left_join(select(db_dataset_df, datasetId = id, name = displayName), .) %>%
  select(-dataset) %>%
  distinct()

# # initial table creation (synIDs need to be integers)
# link_dataset_syntable <- link_dataset_df %>%
#   mutate_at(c("datasetId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("link_dataset", "syn21498902", .) %>%
#   synStore()

# update/overwrite the table
link_dataset_syntable <- update_table("syn21895656", link_dataset_df)
```

```{r}
# add the correct reference/FK for datasetId
description_dataset_df <- description_dataset_df %>%
  inner_join(select(db_dataset_df, datasetId = id, dataset = name), .) %>%
  select(-dataset) %>%
  distinct()

# # initial table creation (synIDs need to be integers)
# description_dataset_syntable <- description_dataset_df %>%
#   mutate_at(c("datasetId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("description_dataset", "syn21498902", .) %>%
#   synStore()

# update/overwrite the table
description_dataset_syntable <- update_table("syn21895659", description_dataset_df)
```

### Tool

#### Minimal Tool annotations

The min-ified `tool` table is intended to only include those attributes/fields that are specific to tools.

```{r}
# remove extraneous columns, clean up column names
min_tool_df <- full_tool_df %>% 
  select(-PublicationTitle, -grantType, -consortium, -pubMedLink, -grantName,
         -publicationTitle, -PubMed, -study, -projectId, -methodID) %>% 
  mutate(id = toolId) %>% 
  select(id, grantId, name = toolId, displayName = toolName, description,
         inputDataType, outputDataType, softwareLanguage, softwareType,
         externalLinkUrl = externalLink) %>% 
  mutate_all(str_trim)
```

#### Description-Tool

```{r}
# pull out descriptions (which exceed column size limits for File Views) 
# to create description-tool association table (one-to-one relationship)
description_tool_df <- min_tool_df %>% 
  select(tool = id, description) %>% 
  filter(!is.na(description)) %>% 
  distinct()

description_tool_df
```

#### Datatype-Tool

```{r}
# pull out input and output datatypes to create datatype-tool association 
# table (one-to-many relationship for tool -> datatypes)
datatype_tool_df <- min_tool_df %>%
  select(id, inputDataType, outputDataType) %>%
  gather(role, dataType, -id) %>%
  mutate(role = str_replace(role, "DataType", "")) %>%
  select(tool = id, dataType, role) %>%
  separate_rows(dataType, sep = ",") %>% 
  mutate(dataType = str_trim(dataType)) %>%
  filter(!is.na(dataType))

datatype_tool_df
```

#### SoftwareLanguage-Tool

```{r}
# pull out software languages to create language-tool association 
# table (one-to-many relationship for tool -> languages)
language_tool_df <- min_tool_df %>% 
  select(tool = id, softwareLanguage) %>% 
  filter(!is.na(softwareLanguage)) %>% 
  separate_rows(softwareLanguage, sep = ",") %>% 
  mutate(softwareLanguage = str_trim(softwareLanguage)) %>%
  mutate(softwareLanguage = case_when(
    str_detect(softwareLanguage, "C++") ~ "C++",
    TRUE ~ softwareLanguage
  ))

language_tool_df
```

#### Link-Tool

```{r}
# pull out links to create link-tool association 
# table (one-to-one relationship for tool -> links)
link_tool_df <- min_tool_df %>% 
  mutate(source = "") %>% 
  select(tool = id, source, url = externalLinkUrl) %>% 
  mutate(source = case_when(
    str_detect(url, "github") ~ "GitHub",
    str_detect(url, "bitbucket") ~ "Bitbucket",
    str_detect(url, "readthedocs") ~ "ReadTheDocs",
    str_detect(url, "bioconductor") ~ "Bioconductor",
    TRUE ~ source
  )) %>%
  mutate(source = ifelse(!is.na(url) & source == "", "Website", source)) %>% 
  filter(source != "")

link_tool_df
```


#### Build/store Tool table (view)

```{r}
tool_df <- min_tool_df %>%
  select(id, grantId, displayName, softwareType) %>% 
  rename(toolType = softwareType) %>% 
  distinct()

# create tool entities in tools folder
tool_entities <- create_entities(tool_df, "syn21645176")

# manually create the view in Synapse and download (NOTE: need to add placeholder
# columns for addtiional annotations as well), then update the view
tool_synview <- update_view("syn21645208", tool_df)

# fetch the tool db table for mapping tool IDs below
db_tool_df <- get_table_df("syn21645208")
```

#### Store association tables

```{r}
# add the correct reference/FK for toolId
description_tool_df <- description_tool_df %>%
  inner_join(select(db_tool_df, toolId = id, tool = name), .) %>%
  select(-tool) %>%
  distinct()

# # initial table creation (synIDs need to be integers)
# description_tool_syntable <- description_tool_df %>%
#   mutate_at(c("toolId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("description_tool", "syn21498902", .) %>%
#   synStore()

# update/overwrite the table
description_tool_syntable <- update_table("syn21918273", description_tool_df)
```

```{r}
# add the correct reference/FK for toolId
datatype_tool_df <- datatype_tool_df %>%
  inner_join(select(db_tool_df, toolId = id, tool = name), .) %>%
  select(-tool) %>%
  distinct()

# # initial table creation (synIDs need to be integers)
# datatype_tool_syntable <- datatype_tool_df %>%
#   mutate_at(c("toolId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("datatype_tool", "syn21498902", .) %>%
#   synStore()

# update/overwrite the table
datatype_tool_syntable <- update_table("syn21683485", datatype_tool_df)
```

```{r}
# add the correct reference/FK for toolId
language_tool_df <- language_tool_df %>%
  inner_join(select(db_tool_df, toolId = id, tool = name), .) %>%
  select(-tool) %>%
  distinct()

# # initial table creation (synIDs need to be integers)
# language_tool_syntable <- language_tool_df %>%
#   mutate_at(c("toolId"), ~ str_replace(., "syn", "")) %>%
#   synBuildTable("language_tool", "syn21498902", .) %>%
#   synStore()

# update/overwrite the table
language_tool_syntable <- update_table("syn21683497", language_tool_df)
```

```{r}
# add the correct reference/FK for toolId
link_tool_df <- link_tool_df %>%
  inner_join(select(db_tool_df, toolId = id, tool = name), .) %>%
  select(-tool) %>%
  distinct()

# # initial table creation (synIDs need to be integers)
link_tool_syntable <- link_tool_df %>%
  mutate_at(c("toolId"), ~ str_replace(., "syn", "")) %>%
  synBuildTable("link_tool", "syn21498902", .) %>%
  synStore()

# update/overwrite the table
link_tool_syntable <- update_table("syn21930556", link_tool_df)
```

---

> Everything below this point is relatively unpolished/experimental...

### File

```{r}
tool_cols <- c(
  "softwareType",
  "inputDataType",
  "outputDataType",
  "softwareAuthor",
  "softwareLanguage",
  "softwareRepository"
)
```

```{r}
experiment_cols <- c(
  "compoundName",
  "compoundDose",
  "experimentalCondition"
)
```

```{r}
assay_cols <- c(
  "assay",
  "experimentalStrategy",
  "nucleicAcidSource",
  "libraryPrep",
  "isStranded",
  "platform",
  "runType"
)
```

```{r}
assay_include_cols <- c(
  "assay",
  "experimentalStrategy",
  "platform"
)
```


```{r}
biospecimen_cols <- c(
  "Title",
  "cellLine",
  "cellSubType",
  "cellType",
  "individualID",
  "individualIdSource",
  "isCellLine",
  "isPrimaryCell",
  "modelSystem",
  "transplantationDonorSpecies",
  "transplantationDonorTissue",
  "transplantationRecipientTissue",
  "transplantationType",
  "diseaseSubtype",
  "diagnosis",
  "tumorType",
  "organ",
  "sex",
  "specimenID",
  "specimenIdSource",
  "tissue",
  "species"
)
```

```{r}
biospecimen_include_cols <- c(
  "Title",
  "tumorType",
  "tissue",
  "species",
  "sex"
)
```

```{r}
method_cols <- c(
  "alignmentMethod",
  "peakCallingMethod",
  "transcriptQuantificationMethod"
)
```

```{r}
min_datafile_df <- full_file_df %>% 
  filter(is.na(softwareType)) %>%
  select(-tool_cols) %>% 
  # select(-experiment_cols) %>%
  # select(-biospecimen_cols) %>%
  # select(-method_cols) %>%
  select_if(function(x){ !all(is.na(x)) }) %>% 
  select(-consortium, -fundingAgency, -modifiedOn, -createdOn, -dataset,
         -study_name, -centerName, -grantType, -resourceType,
         -PIInstitution, -analysisType, -concreteType, -study, -studies,
         -principalInvestigator, -networkEdgeType, -Theme) %>% 
  select(id, grantId = projectId, parentId, datasetId = parentId, name, 
         dataType, dataSubtype, fileFormat, isMultiIndividual, isMultiSpecimen,
         externalLink = geo_accession, everything()) %>% 
  mutate_all(str_trim) %>% 
  filter(!grantId %in% c("syn7248578"))
```

```{r}
assay_datafile_df <- min_datafile_df %>% 
  select(id, dataType, one_of(assay_include_cols)) %>% 
  mutate(dataType = case_when(
    assay == "rnaArray" ~ "geneExpression",
    assay == "rnaSeq" ~ "geneExpression",
    assay == "ATACSeq" ~ "chromatinActivity",
    experimentalStrategy == "Whole Transcriptome Sequencing" ~ "geneExpression",
    assay == "migration" ~ "?",
    assay == "widefieldFluorescenceMicroscopy" ~ "?",
    assay == "tractionForceMicroscopy" ~ "?",
    assay == "simulation" ~ "?",
    assay == "cellViabilityAssay" ~ "drugScreen",
    assay == "methylationArray" ~ "chromatinActivity", 
    TRUE ~ dataType
  )) %>% 
    mutate(assay = case_when(
    experimentalStrategy == "Whole Transcriptome Sequencing" ~ "rnaSeq",
    TRUE ~ assay
  )) %>% 
  mutate(experimentalStrategy = case_when(
    assay == "rnaArray" ~ "cDNA Array",
    assay == "rnaSeq" ~ "Whole Transcriptome Sequencing",
    assay == "ATACSeq" ~ "ATAC-Seq",
    assay == "bisulfiteSeq" ~ "Bisulfite Sequencing",
    assay == "RPPA" ~ "Ribosomal P Protein Antibody Measurement",
    assay == "migration" ~ "Cell Migration Assay",
    TRUE ~ experimentalStrategy
  )) %>% 
  # select(-id) %>%
  filter(!experimentalStrategy == "NA") %>% 
  select(id, assay = experimentalStrategy, platform) %>%
  distinct()

assay_datafile_df
```

```{r}
biospecimen_datafile_df <- min_datafile_df %>% 
  select(id, dataType, one_of(biospecimen_include_cols)) %>% 
  mutate(Title = ifelse(Title == "NA", NA, Title)) %>% 
  mutate(tissue = case_when(
    tissue == "Embyro" ~ "Embryo",
    TRUE ~ tissue
  )) %>% 
  select(-Title, -dataType) %>% 
  # select(individualID, specimenID) %>% 
  distinct()

biospecimen_datafile_df
```

### Team

```{r}
min_team_df <- min_grant_df %>% 
  distinct(synapseTeam) %>% 
  filter(!is.na(synapseTeam)) %>% 
  rename(id = synapseTeam) %>% 
  mutate(name = map_chr(id, ~ synGetTeam(.)$name))
```

```{r}
team_syntable <- synBuildTable("team", "syn21498902", min_team_df)
team_syntable <- synStore(team_syntable)
```

### User

```{r}
as_user_df <- function(user_obj) {
  user_obj$member$json() %>% 
    jsonlite::fromJSON() %>% 
    as_tibble() %>% 
    mutate(isAdmin = user_obj$isAdmin,
           teamId = user_obj$teamId)
}

flatten_user_dfr <- function(team_member_list) {
  team_member_list %>% 
    map_dfr(as_user_df)
}

person_team_df <- min_team_df$id %>% 
  map_dfr(function(team_id) {
    synGetTeamMembers(team_id)$asList() %>% 
      flatten_user_dfr()
  })
```


```{r}
person_team_df %>% 
  select(-isAdmin, -teamId) %>% 
  distinct() %>% 
  arrange(lastName)
```